Fix `FW_COMMIT_HASH` compiler warning

gudnimg opened the pull request:
On my end, the default length for the abbreviated commit hash is 9 characters and won't fit into `uint32_t`.

To fix this I propose to change `FW_COMMIT_HASH` into a string and create preprocessor symbol for the string length so that it's known at compile time.

If the string should be longer or shorter then only `FW_COMMIT_HASH_LENGTH` needs to be configured on the CMake side

- [ ] Cherry-pick PR to MK3_3.13.2

gudnimg committed:
Fix FW_COMMIT_HASH compiler warning

On my end, the default length for the abbreviated commit hash is 9 characters.
This won't fit into uint32_t (4 bytes).

Instead change FW_COMMIT_HASH into a string
and create preprocessor symbol for the string length
such that it's known at compile time.

If the string should be longer or shorter
then only FW_COMMIT_HASH_LENGTH needs to be configured on the CMake side

gudnimg commented:
The function `FW_VERSION_HASH_P()` isn't called anywhere yet. Is it supposed to be in M115?

gudnimg committed:
Fix two python commands on Windows

vintagepc commented on cmake/ProjectVersion.cmake:
```diff
@@ -87,18 +87,30 @@ ENDIF()
+set(FW_COMMIT_HASH_LENGTH 9)
+
+# Create fallback value with constant length
+string(REPEAT "0" ${FW_COMMIT_HASH_LENGTH} FW_COMMIT_HASH_UNKNOWN)
```
This is at odds with/will get clobbered by line 94-ish of CMakeLists.txt.

gudnimg replied:
You're right, I missed it. Will look into it.

gudnimg committed:
Use FW_COMMIT_HASH_UNKNOWN in CMakeLists.txt

gudnimg commented:
@vintagepc the last commit seems to have resolved the static assert :) 

There is still an issue with the Github Action. The runner doesn't detect `git`. It's unrelated to this PR but I don't see why this is happening. The Ubuntu version we're using should have Git included according to Github.

I've tried to run the Workflow via Docker on Windows using https://github.com/nektos/act but no luck so far.

![image](https://github.com/prusa3d/Prusa-Firmware/assets/8218499/01cf94ae-b351-4e6a-a679-2c4ecc4a660b)

gudnimg committed:
Fix git_describe_working_tree failing

The command
/usr/bin/git describe --abbrev=0 --dirty=-D --broken=-B

Was returning error code 128 and no output.

In the workflow file, specifying fetch-depth = 0, will make sure
to fetch all branches and tags. This seems to fix the issue.

gudnimg commented:
Fixed the issue with the git description command. The problem was not that git command wasn't found, our script is just interpreting the error that way :) 
![image](https://github.com/prusa3d/Prusa-Firmware/assets/8218499/098edcb7-0b7c-4d0c-ad03-aec48d332822)

3d-gussner commented:
We can shorten the hash to 8 chars with `--abbrev=8` in https://github.com/gudnimg/Prusa-Firmware/blob/fix-compiler-warning-sha/cmake/GetGitRevisionDescription.cmake#L353.

gudnimg commented:
> We can shorten the hash to 8 chars with `--abbrev=8` in https://github.com/gudnimg/Prusa-Firmware/blob/fix-compiler-warning-sha/cmake/GetGitRevisionDescription.cmake#L353.

Note that `git_head_commit_data(FW_COMMIT_DATE "%ct")` is using the same function. We would need to create a new function.

I think having the data as a string (`char *`) in the firmware is better than `uint32_t`, because then we don't need to convert the `uint32_t` to hexidecimal at runtime.

3d-gussner commented:
Okay. Is the other comment resolved?

gudnimg commented:
> Okay. Is the other comment resolved?

Yeah https://github.com/prusa3d/Prusa-Firmware/pull/4432/commits/69e06bac8416f620763b2499c2cffcb94c35c0a5 resolved it. Github always failed to build before that fix because I added a new `static_assert` which checks the string length is correct.

3d-gussner reviewed (approved):
Tested and works well. No warnings.

3d-gussner commented:
@vintagepc Can we merge it?
